### PR TITLE
[CALCITE-3921] Support TableModify json serialization and deserialization

### PR DESCRIPTION
**Pull request by yanlin-Lynn**
Currently, trying to serialize a TableModify to json string will cause exception, also for trying to deserialize from json string. See jira [CALCITE-3912](https://issues.apache.org/jira/browse/CALCITE-3921) for full stacktrace.
This PR tries to support serialize TableModify to json string and deserialize from it.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java [hunk 1]
```diff
@@ -199,6 +199,9 @@ public boolean getBoolean(String tag, boolean default_) {
       public List<RexNode> getExpressionList(String tag) {
         @SuppressWarnings("unchecked")
         final List<Object> jsonNodes = (List) jsonRel.get(tag);
+        if (jsonNodes == null) {
+          return null;
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java:203]: When the `jsonNodes` could be null ?

Comment (yanlin-Lynn) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java:203]: when serialize TableModify, the 'updateColumnList' and 'sourceExpressionList' may be null. 
For exampl, for this relnode tree
```
LogicalTableModify(table=[[scott, EMP]], operation=[INSERT], flattened=[false])
  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
    LogicalTableScan(table=[[scott, EMP]])
```
the serialized json string is
```
{
  "rels": [
    {
      "id": "0",
      "relOp": "LogicalTableScan",
      "table": [
        "scott",
        "EMP"
      ],
      "inputs": []
    },
    {
      "id": "1",
      "relOp": "LogicalProject",
      "fields": [
        "EMPNO",
        "ENAME",
        "JOB",
        "MGR",
        "HIREDATE",
        "SAL",
        "COMM",
        "DEPTNO"
      ],
      "exprs": [
        {
          "input": 0,
          "name": "$0"
        },
        {
          "input": 1,
          "name": "$1"
        },
        {
          "input": 2,
          "name": "$2"
        },
        {
          "input": 3,
          "name": "$3"
        },
        {
          "input": 4,
          "name": "$4"
        },
        {
          "input": 5,
          "name": "$5"
        },
        {
          "input": 6,
          "name": "$6"
        },
        {
          "input": 7,
          "name": "$7"
        }
      ]
    },
    {
      "id": "2",
      "relOp": "LogicalTableModify",
      "table": [
        "scott",
        "EMP"
      ],
      "operation": "INSERT",
      "flattened": false
    }
  ]
}
```
When trying to get updateColumnList by `input.getExpressionList("sourceExpressionList")`, the jsonNodes will be null.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java [hunk 2]
```diff
@@ -895,6 +898,38 @@ private RelNode mockCountOver(String table,
...
+    final RelBuilder builder = RelBuilder.create(config);
+    final RexBuilder rexBuilder = builder.getRexBuilder();
+    LogicalTableScan scan = (LogicalTableScan) builder.scan("EMP").build();
+    List<RexNode> projExprs = new ArrayList<>();
+    int inputFieldNum = scan.getRowType().getFieldCount();
+    for (int i = 0; i < inputFieldNum; i++) {
+      RexNode projExpr = rexBuilder.makeInputRef(scan, i);
+      projExprs.add(projExpr);
+    }
+    LogicalProject project = LogicalProject.create(
```
Comment (vineetgarg02) [inline core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:912]: May be use `builder` here too?

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java:203]: Well, the `targetColumnNameList` is always null for operation `INSERT` and `DELETE`;
the `sourceExpressionList` is always null for `INSERT`, `DELETE` and `MERGE`.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java [hunk 3]
```diff
@@ -895,6 +898,38 @@ private RelNode mockCountOver(String table,
...
+        scan, ImmutableList.of(), projExprs, scan.getRowType());
+    LogicalTableModify modify = LogicalTableModify.create(
+        scan.getTable(),
+        (Prepare.CatalogReader) scan.getTable().getRelOptSchema(),
+        project,
+        TableModify.Operation.INSERT,
+        null,
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:919]: All the operations should be tested, i.e. the `DELETE`, `UPDATE` and `MERGE`.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java (hunk 2 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:912]: Yes, we should.

Comment (chunweilei) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:912]: +1

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java (hunk 3 above)
Comment (yanlin-Lynn) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:913]: yes, I agree.

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java (hunk 1 above)
Comment (yanlin-Lynn) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJsonReader.java:203]: yes, I update the null check logic in constructor.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java (hunk 2 above)
Comment (yanlin-Lynn) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java]: Yes, that's better.

**Comment by yanlin-Lynn** [comment]
Thanks for reviewing and merging.

**Pull request closed without merging**